persist/sqlite/revert.go

```diff
@@ -2,53 +2,242 @@ package sqlite
+	return clause
+}
+
+func deleteEvents(tx *txn, bid types.BlockID) error {
+	rows, err := tx.Query(`SELECT id FROM events WHERE block_id = ?`, encode(bid))
+	if err != nil {
+		return fmt.Errorf("failed to query event IDs: %w", err)
+	}
+	defer rows.Close()
+
+	var dbIDs []int64
+	for rows.Next() {
+		var dbID int64
+		if err := rows.Scan(&dbID); err != nil {
+			return fmt.Errorf("failed to scan event ID: %w", err)
+		}
+		dbIDs = append(dbIDs, dbID)
+	}
+	if err := rows.Err(); err != nil {
+		return fmt.Errorf("failed to get event ID rows: %w", err)
+	}
+
+	clause := generateInClause(dbIDs)
+	if _, err := tx.Exec(`DELETE FROM event_addresses WHERE event_id IN` + clause); err != nil {
+		return fmt.Errorf("failed to delete from event_addresses table: %w", err)
+	} else if _, err := tx.Exec(`DELETE FROM v1_transaction_events WHERE event_id IN` + clause); err != nil {
+		return fmt.Errorf("failed to delete from v1_transaction_events table: %w", err)
+	} else if _, err := tx.Exec(`DELETE FROM v2_transaction_events WHERE event_id IN` + clause); err != nil {
+		return fmt.Errorf("failed to delete from v2_transaction_events table: %w", err)
+	} else if _, err := tx.Exec(`DELETE FROM payout_events WHERE event_id IN` + clause); err != nil {
+		return fmt.Errorf("failed to delete from payout_events table: %w", err)
+	} else if _, err := tx.Exec(`DELETE FROM v1_contract_resolution_events WHERE event_id IN` + clause); err != nil {
+		return fmt.Errorf("failed to delete from v1_contract_resolution_events table: %w", err)
+	} else if _, err := tx.Exec(`DELETE FROM v2_contract_resolution_events WHERE event_id IN` + clause); err != nil {
+		return fmt.Errorf("failed to delete from v2_contract_resolution_events table: %w", err)
+	} else if _, err := tx.Exec(`DELETE FROM events WHERE id IN` + clause); err != nil {
+		return fmt.Errorf("failed to delete from events table: %w", err)
+	}
+	return nil
+}
+
+func deleteLastContractRevisions(tx *txn, bid types.BlockID) error {
+	if _, err := tx.Exec(`DELETE FROM last_contract_revision WHERE confirmation_block_id = ?;`, encode(bid)); err != nil {
+		return fmt.Errorf("failed to delete from last_contract_revision: %w", err)
+	} else if _, err := tx.Exec(`DELETE FROM v2_last_contract_revision WHERE confirmation_block_id = ?;`, encode(bid)); err != nil {
+		return fmt.Errorf("failed to delete from v2_last_contract_revision: %w", err)
+	}
+	return nil
+}
+
 // deleteV1Transactions deletes the transactions from the database if they are
 // not referenced in any blocks.
-func deleteV1Transactions(tx *txn, txns []types.Transaction) error {
-	stmt, err := tx.Prepare(`DELETE FROM transactions AS t
-WHERE t.transaction_id = ?
-	AND NOT EXISTS (
-		SELECT 1
-		FROM block_transactions bt
-		WHERE bt.transaction_id = t.id
-);`)
+func deleteV1Transactions(tx *txn, bid types.BlockID) error {
+	rows, err := tx.Query(`SELECT transaction_id AS id
+FROM
+	block_transactions
+WHERE
+	block_id = ?
+	AND transaction_id NOT IN (
```

Comment:
I wonder if it's noticeably faster to `GROUP BY` transaction_id and have a distint count on block_id that equals 1. So something like

```sql
SELECT transaction_id
FROM block_transactions
GROUP BY transaction_id
HAVING COUNT(DISTINCT block_id) = 1 AND ANY(block_id) = ?
```

Reply:
Thanks for checking